pytket/extensions/cutensornet/mps/mps.py

```diff
@@ -407,32 +413,340 @@ def vdot(self, other: MPS) -> complex:
             raise RuntimeError(
                 "The qubit labels or their position on the MPS do not match."
             )
+        if len(self) == 0:
+            raise RuntimeError("There are no tensors in the MPS.")
 
         self._flush()
         other._flush()
 
-        # The two MPS will be contracted from left to right, storing the
-        # ``partial_result`` tensor.
-        partial_result = cq.contract(
-            "LRp,lrp->Rr", self.tensors[0].conj(), other.tensors[0]
-        )
-        # Contract all tensors in the middle
-        for pos in range(1, len(self) - 1):
+        # Special case if only one tensor remains
+        if len(self) == 1:
+            result = cq.contract("LRp,lrp->", self.tensors[0].conj(), other.tensors[0])
+
+        else:
+            # The two MPS will be contracted from left to right, storing the
+            # ``partial_result`` tensor.
             partial_result = cq.contract(
-                "Ll,LRp,lrp->Rr",
+                "LRp,lrp->Rr", self.tensors[0].conj(), other.tensors[0]
+            )
+            # Contract all tensors in the middle
+            for pos in range(1, len(self) - 1):
+                partial_result = cq.contract(
+                    "Ll,LRp,lrp->Rr",
+                    partial_result,
+                    self.tensors[pos].conj(),
+                    other.tensors[pos],
+                )
+            # Finally, contract the last tensor
+            result = cq.contract(
+                "Ll,LRp,lrp->",  # R and r are dim 1, so they are omitted; scalar result
                 partial_result,
+                self.tensors[-1].conj(),
+                other.tensors[-1],
+            )
+
+        return complex(result)
+
+    def sample(self) -> dict[Qubit, int]:
+        """Returns a sample from a Z measurement applied on every qubit.
+
+        Notes:
+            The MPS ``self`` is not updated. This is equivalent to applying
+            ``mps = self.copy()`` then ``mps.measure(mps.get_qubits())``.
+
+        Returns:
+            A dictionary mapping each of the qubits in the MPS to their 0 or 1 outcome.
+        """
+        mps = self.copy()

```

Comment:
Hmm... yeah, now that you mention it, I could write an alternative version of the algorithm currently in `measure` so that the MPS is not consumed, then copying would no longer be required. However, that'd require some code duplication... I'm sure I could do some refactoring so that code is not duplicated, but it's not immediate. I'll leave it in the TODO list for now.